src/common/mfu_flist_copy.c

```diff
@@ -92,67 +100,78 @@ static mfu_copy_stats_t mfu_copy_stats;
 static mfu_copy_file_cache_t mfu_copy_src_cache;
 static mfu_copy_file_cache_t mfu_copy_dst_cache;
 
-static int mfu_copy_open_file(const char* file, int read_flag,
-        mfu_copy_file_cache_t* cache, mfu_copy_opts_t* mfu_copy_opts)
+static void mfu_copy_open_file(const char* file, int read_flag,
+        mfu_copy_file_cache_t* cache, mfu_copy_opts_t* mfu_copy_opts,
+        mfu_file_t* mfu_file)
 {
-    int newfd = -1;
-
     /* see if we have a cached file descriptor */
     char* name = cache->name;
     if (name != NULL) {
         /* we have a cached file descriptor */
-        int fd = cache->fd;
+        int fd         = cache->fd;
+#ifdef DAOS_SUPPORT
+        dfs_obj_t* obj = cache->obj;
+#endif
         if (strcmp(name, file) == 0 && cache->read == read_flag) {
             /* the file we're trying to open matches name and read/write mode,
              * so just return the cached descriptor */
-            return fd;
+            if (fd > 0) {
+                return fd;
+            } 
+#ifdef DAOS_SUPPORT
+            else {
+                return obj;
+            }
+#endif
         } else {
             /* the file we're trying to open is different,
              * close the old file and delete the name */
-            mfu_close(name, fd);
+            mfu_file_close(name, mfu_file);
             mfu_free(&cache->name);
         }
     }
-
     /* open the new file */
     if (read_flag) {
         int flags = O_RDONLY;
         if (mfu_copy_opts->synchronous) {
             flags |= O_DIRECT;
         }
-        newfd = mfu_open(file, flags);
+        /* set obj or fd here */
+        mfu_file_open(file, flags, mfu_file);
     } else {
         int flags = O_WRONLY | O_CREAT;
         if (mfu_copy_opts->synchronous) {
             flags |= O_DIRECT;
         }
-        newfd = mfu_open(file, flags, DCOPY_DEF_PERMS_FILE);
+        /* set obj or fd here */
+        mfu_file_open(file, flags, mfu_file, DCOPY_DEF_PERMS_FILE);
     }
 
     /* cache the file descriptor */
-    if (newfd != -1) {
+    if (mfu_file->fd != -1) {
         cache->name = MFU_STRDUP(file);
-        cache->fd   = newfd;
+        cache->fd   = mfu_file->fd;
         cache->read = read_flag;
+#ifdef DAOS_SUPPORT
+        cache->obj  = mfu_file->obj;
+#endif
 #ifdef LUSTRE_SUPPORT
```

Comment:
We should do some testing in an environment that has both Lustre and DAOS enabled.  For example, this code assumes the fd field is valid, but if it's dealing with a DAOS object, maybe that's not the case.  This particular code segment might be protected since the group lock is a lustre feature, but there might be other code like this out there.

Are you running on a system where you can have both enabled at once?

Reply:
@adammoody No, I don't have an environment currently where I can run both unfortunately. The fd field should be initialized to -1 from what I saw, and as long as there is checking it shouldn't be an issue? Because if DAOS is enabled, the fd field will continue to stay -1?